### PR TITLE
feat(core,pubsub): cross-process agent signal coordination via lease providers + unified accepted result

### PR DESCRIPTION
**Pull request by CalebBarnes**
## Summary

Two paired primitives that let multiple processes coordinate around a single agent thread, plus a unified `accepted` return on the signal/message APIs that reports what the runtime actually did with a signal.

### 1. `LeaseProvider` capability

Distributed leasing is a separate concern from event delivery, and not every pubsub backend can coordinate a lock (GCP PubSub's "lease" is an ack-deadline, not a mutex; Kafka has no native lock). So leasing is **not** on the `PubSub` base class — it's a standalone capability that backends opt into.

- **`LeaseProvider` interface** — `acquireLease`, `releaseLease`, `renewLease`, `getLeaseOwner`, and `transferLease`. Lets callers elect a single owner for a key and atomically hand ownership to a new owner without ever freeing the key.
- **`isLeaseProvider(value)`** — structural (duck-type) guard, so detection works across package boundaries without `instanceof` fragility.
- **`NoopLeaseProvider`** — always-win / no-op fallback that preserves single-process semantics (every caller wins its own lease race).
- `EventEmitterPubSub implements LeaseProvider`: in-memory `Map` with TTL.
- `RedisStreamsPubSub implements LeaseProvider`: `SET key value NX PX ttl` to atomically claim, Lua `EVAL` compare-and-delete on release, Lua-guarded TTL extension on renew, Lua-guarded `GET == fromOwner → SET toOwner PX ttl` for `transferLease`, plus `getLeaseOwner`. `transferLease` on `NoopLeaseProvider` is an always-win no-op; providers without it degrade to release-then-acquire.
- `CachingPubSub` is transparent to leasing — it surfaces its inner backend's lease provider via `getLeaseProvider()` rather than faking the capability.
- The signals runtime feature-detects the provider **once** at the boundary (`#getLeaseProvider`) and routes all lease call sites through it, falling back to `NoopLeaseProvider` when the configured pubsub can't lease. Backends with no native lock (e.g. a plain `PubSub`) therefore keep today's always-win behavior with no per-call-site `instanceof` checks.

### 2. `accepted` on the signal/message result

`sendSignal` (and the message/state/notification variants) now return a single `accepted` promise reporting what the runtime did with the signal. This replaces the previous synchronous `accepted: true` flag, the separate `outcome` promise, and the best-effort top-level `runId` — all collapsed into one promise that resolves at the routing-decision point:

```ts
accepted: Promise<SendAgentSignalAccepted<OUTPUT>>;

type SendAgentSignalAccepted<OUTPUT = unknown> =
  | { action: 'wake'; runId: string; output: MastraModelOutput<OUTPUT> }
  | { action: 'deliver'; runId: string }
  | { action: 'persist' }
  | { action: 'discard' };
```

`action` mirrors the input `behavior` verbs. When the promise resolves depends on the action — in every case it is well before any woken run *finishes*:

| `action` | meaning | `accepted` resolves when |
| --- | --- | --- |
| `wake` | this runtime started the run; `runId` identifies it and `output` is its `MastraModelOutput` for in-process consumption | the run has started locally (the lease was won) |
| `deliver` | the signal was queued onto an in-flight run — a follow-up signal joining an active run, or the loser of a cross-process wake race (forwarded to the winning run); `runId` is the run it joined | the signal is enqueued |
| `persist` | a `persist` behavior routed the signal to memory; no run exists, so no `runId`. Durability is reported separately by `persisted` | the persist routing decision is made (the memory write itself is awaited via `persisted`, not `accepted`) |
| `discard` | policy dropped the signal; nothing ran, nothing stored, no `runId` | the discard decision is made |

- **Asymmetric `runId`**: `runId` is present only on `wake`/`deliver`, where a run actually exists. `persist`/`discard` carry just `action` — there is no run to report, and the stored signal is correlated via `result.signal.id`. Callers narrow with `'runId' in accepted` (or `accepted.action === 'wake'`) rather than reading a maybe-undefined field.
- `output` is only present on `'wake'`, and `'wake'` only resolves on the runtime that actually runs the agent — this is what makes `accepted` lease-aware. Opt into awaiting run output via `action === 'wake'` + `output`.
- **Routing decisions never reject** `accepted`; it resolves with the action above. Generation errors surface on the woken run's `output.consumeStream()`, not by rejecting `accepted`. `accepted` *does* reject on setup/misconfiguration failures (no/unsupported model, FGA denial, bad request context) — an honest signal that the signal could not be routed or started. A detached no-op `.catch()` is attached so this rejection never surfaces as an unhandled rejection for callers that don't await it.
- `persisted?: Promise<void>` is kept top-level (existing consumers: notifications dispatcher + tool) and is the durability boundary, distinct from the routing-decision boundary `accepted`. For a `persist` signal, `accepted` resolves immediately at the routing decision while `persisted` resolves later, once the memory write actually completes — callers that need to know the write landed must await `persisted`, not `accepted`.

This is the field the heartbeat worker reads to derive trigger outcomes without re-deriving runtime policy.

---

### Lease winner/loser wake behavior (thread runtime + types)

**Agent runtime lease winner vs loser** (`packages/core/src/agent/thread-stream-runtime.ts`)
- In the idle wake path, the runtime resolves a `LeaseProvider` (via `#getLeaseProvider`) and attempts `lease.acquireLease(threadKey, runId, ttl)`:
  - **Winner**: starts background lease renewal (TTL/3), runs `agent.stream(...)`, resolves `accepted` to `{ action: 'wake', runId, output }`.
  - **Loser**: rolls back the local optimistic "active thread" reservation, publishes `signal-enqueued` (via `#publishAndWait` so the message is flushed before resolution) so the winner picks up the signal, and resolves `accepted` to `{ action: 'deliver', runId }` (the winning run's id).
- Lease acquisition fails open: if the lease backend is unreachable the call is treated as acquired so the caller still gets a response (availability over strict exclusivity, bounded by the lease TTL).
- When the configured pubsub is not a `LeaseProvider`, the runtime resolves `NoopLeaseProvider` (always-win / no-op), which is exactly the single-process behavior the `PubSub` base class previously inherited.

**Lease lifecycle**
- TTL: **15s**, renewal interval **5s** (TTL/3; tolerates a couple of missed renewals). Both are overridable for tests via `MASTRA_AGENT_THREAD_LEASE_TTL_MS` / `MASTRA_AGENT_THREAD_LEASE_RENEW_INTERVAL_MS` (production defaults unchanged).
- Renewal timer `unref()`'d to avoid keeping the process alive.
- Cleanup on `abortRun`, finalizers (`waitUntilFinished`), drain completion, and error/stream-failure paths.
- Crash recovery relies on lease expiry when renewal stops.

**Lease held across the drain chain (cross-process race fix)**
- When a thread run completes, the runtime may still have queued follow-up work for that thread (pending signals, idle signals, continuations). Previously the lease was **released** at run completion and then those drain paths started the next run **without re-acquiring** it. That left the lease key momentarily empty, so a fresh process could win the freed lease and start a competing run for the **same thread** — two concurrent runs.
- Fix: the owner now keeps the lease across the entire drain chain instead of releasing between runs. On completion-with-queued-work it **transfers** the lease to the next run's id via `transferLease` (atomic owner-verified handoff) rather than release-then-acquire, so the key is never empty while work is queued. The lease is only **released** once all queues are empty.
- Every drain path (`#drainPendingSignals`, `#drainPendingIdleSignals`, continuations) now acquires-or-transfers the lease before streaming. If it can't obtain the lease (e.g. a subscriber-side drain on a process that never owned it), it **forwards** the pending signal to the current owner via `signal-enqueued` instead of starting a local run — so non-owning processes can't start competing runs either.
- Backends without `transferLease` degrade to release-then-acquire (a small gap remains only for those; Redis is the only cross-process implementor today and is gap-free).

**Typed API + generic propagation**
- `SendAgentSignalResult`, `SendAgentMessageResult`, `QueueAgentMessageResult`, `SendAgentStateSignalResult`, and `SendAgentNotificationSignalResult` carry `OUTPUT` generics and the new `accepted` field.
- `OUTPUT` is propagated through `sendSignal`, `sendMessage`, `queueMessage`, `sendStateSignal`, and the notification overloads in `packages/core/src/agent/agent.ts`.
- `SendAgentNotificationSignalResult` keeps its own boolean `accepted` (a dispatch flag, distinct from the signal-routing promise) and re-exposes the routing promise as `outcome?: Promise<SendAgentSignalAccepted<OUTPUT>>`.

---

### Tests + Validation

- **EventEmitter lease tests** — acquire/reject, idempotent re-acquire, TTL expiry, release correctness, renew success/failure.
- **`CachingPubSub` lease passthrough tests** — `getLeaseProvider()` returns the inner provider when the inner is a `LeaseProvider`, and `undefined` when it isn't (so the runtime falls back to `NoopLeaseProvider`).
- **Redis lease tests** — contention (single winner), idempotent acquire, TTL takeover, release ownership checks, renewal edge cases, `getLeaseOwner`.
- **Cross-runtime wake race test (core)** — races two runtimes waking the same idle thread via a shared `EventEmitterPubSub`; asserts exactly one winner resolves `accepted` to `wake` + defined `output` while the loser resolves `deliver`, and total underlying `agent.stream` calls is `1`.
- **Signal-routing error mapping (server)** — a rejected `accepted` from a USER-category `MastraError` maps to HTTP 400; any other rejection maps to 500.
- **Redis lease transfer tests** — `transferLease` success by owner, rejection when caller isn't the owner, rejection on expired/missing lease, and full-TTL reset on transfer.
- **Redis cross-process integration** — worker fixture + spawner (`agent-thread-cross-process.test.ts`, conditional). Includes a cross-process lease-race regression test that polls the **raw Redis lease key** (unmaskable by local subscriber state) to prove the key never goes empty during the drain chain: a drained follow-up run holds the lease and a racing process cannot win a competing run.

---

### Documentation

- **Release notes** (`.changeset/pubsub-lease-owner-stream.md`) — lease/ownership model and `accepted` semantics.
- **Agent reference docs** (`docs/src/content/en/reference/agents/agent.mdx`) — documents `accepted` on `sendMessage`, `queueMessage`, `sendSignal`, `sendStateSignal`, and `sendNotificationSignal`, including the serverless `consumeStream()` keep-alive example.

---

## Validation Performed

- ✅ `pnpm build:core`
- ✅ `pnpm --filter ./packages/core check`
- ✅ `pnpm --filter ./packages/core lint`
- ✅ Focused Vitest (`agent-signals` + `event-emitter` + server handler signal-routing)
- ✅ `pnpm --filter @mastra/redis-streams build`
- ✅ `pnpm --filter @mastra/redis-streams lint`
- ✅ docs prettier + remark clean

## ELI5

Imagine you have multiple workers running in different processes, and they all want to wake up and run the same agent (like a shared job queue). Instead of them all trying to run it simultaneously, this PR lets them compete for a "lease" (a temporary lock). The winner gets the lease, runs the agent, and processes all the signals. The losers get told "I'm forwarding your request to the winner" instead of trying to run it themselves. Plus, instead of getting a simple "yes/no" answer about whether your signal was accepted, you now get a promise that explains exactly what happened: did the agent start here, was your signal forwarded to another process, or was it saved for later?

---

## Summary

This PR introduces **distributed lease coordination** and **unified signal routing results** to enable multiple processes to safely coordinate around a single agent thread.

### LeaseProvider Interface

A new distributed leasing capability (`LeaseProvider`) separates lease management from event delivery. PubSub backends optionally implement leasing via four methods:
- `acquireLease(key, owner, ttlMs)` — grants ownership with TTL; idempotent for same owner
- `getLeaseOwner(key)` — reads current owner or `undefined`
- `releaseLease(key, owner)` — removes lease if caller is owner
- `renewLease(key, owner, ttlMs)` — extends TTL for current owner only

**Implementations:**
- `EventEmitterPubSub` — in-memory `Map` with TTL tracking
- `RedisStreamsPubSub` — atomic `SET NX PX` acquire, Lua compare-and-delete release, Lua-guarded renewal
- `CachingPubSub` — transparent pass-through to inner backend's `LeaseProvider` (or `undefined`)
- `NoopLeaseProvider` — fallback for single-process; always-win, preserving backward compatibility

Runtime detects provider once at startup and routes all lease calls through it.

### Unified `accepted` Result

`sendSignal` and variants now return `accepted: Promise<SendAgentSignalAccepted<OUTPUT>>` describing the routing decision instead of synchronous boolean + separate outcome promise:
- `'wake'` — this process won the lease and started the agent; includes `runId` and `output` stream
- `'deliver'` — signal forwarded to in-flight run (lease loser); includes `runId` of winner's run
- `'persist'` — queued to storage for later; no `runId`
- `'discard'` — policy filtered it out; no `runId`

`accepted` always resolves (never rejects) on routing decisions; generation/handling errors for `wake` surface via `output.consumeStream()`. Rejects only on setup failures (no model, FGA denial, bad context).

### Cross-Process Wake Coordination

When `ifIdle.behavior` is `wake` and multiple processes race:
1. **Lease acquire** — first process acquires distributed lease (TTL: 15s, renewal: 5s intervals)
2. **Lease winner** — starts lease renewal background timer, runs `agent.stream()`, resolves `accepted` to `{ action: 'wake', runId, output }`
3. **Lease loser** — rolls back local reservation, publishes `signal-enqueued` event to winner, resolves `accepted` to `{ action: 'deliver', runId: winner's }` so caller skips local consumption
4. **Lease release** — winner releases lease on stream completion/abort/error; loser publishes signal to durability queue

Lease acquisition "fails open" on backend errors (treat as acquired for availability).

### Type Updates

- `SendAgentSignalResult<OUTPUT>` — now generic with `accepted: Promise<SendAgentSignalAccepted<OUTPUT>>`
- `SendAgentMessageResult<OUTPUT>` / `QueueAgentMessageResult<OUTPUT>` — generic aliases forwarding to signal result
- `SendAgentStateSignalResult<OUTPUT>` — generic; skipped variant no longer carries `accepted`/`runId`
- `SendAgentNotificationSignalResult<OUTPUT>` — optional `accepted` for underlying signal when notification emits one

### Implementation Details

- **Thread runtime** — `AgentThreadStreamRuntime` tracks per-`runId` renewal timers (`leaseRenewalTimers`), acquires lease before idle wake, releases on prepared-run abort/persisted-run finish/watched-run finish/stream error
- **Test coverage** — New lease tests in `event-emitter.test.ts`, `pubsub.test.ts` (Redis), `caching-pubsub.test.ts`, and a skipped Redis integration test spawning three child processes to verify cross-process lease contention and signal durability
- **Documentation** — `agent.mdx` updated with promise-based `accepted` semantics and serverless `waitUntil` example
- **Consumer updates** — Signal tests, harness, notifications dispatcher, and server handlers updated to await `accepted` and extract `runId` from settled action

### Known Blocking Issues

1. **Linting failures in `@mastra/redis-streams`** — Two import-style violations require correction:
   - `agent-thread-cross-process.test.ts:19` — `ManagedProcess` must use `import type`
   - `agent-thread-worker.entry.ts:23` — `Agent` must use `import type`

2. **PR description clarity** — `AgentChannels` integration (skipping `consumeStream()` when `ownerStream` is `undefined`) is deferred to follow-up PRs; description should clarify this scope boundary

3. **Message flush timing** — Lease-loser path publishes `signal-enqueued` fire-and-forget; serverless callers using `ownerStream` resolution as "safe to exit" boundary may lose the message on abrupt process termination. Consider using `publishAndWait()` to ensure durability before resolving loser's promise

### Testing

- **Core validation:** 101 focused Vitest tests passed (`agent-signals` + `event-emitter`)
- **Build:** `pnpm build:core` and `pnpm --filter `@mastra/redis-streams` build` passed
- **Lint:** `@mastra/redis-streams lint` **failed** (blocking)
- **Redis integration:** Conditional; Docker unavailable in test environment

### Deferred Work

`AgentChannels` consumer integration is intentionally stacked as a follow-up PR. Core primitives and runtime integration ship in this PR.

### CODE REVIEW

**Comment by rphansen91** [comment]
@coderabbitai review

**Comment by TylerBarnes** [comment]
Thanks for the detailed PR. The overall implementation direction looks good: the lease API lives at the PubSub boundary, Redis uses atomic owner-checked operations, and the thread runtime only starts `agent.stream()` for the lease winner. The focused core coverage passed locally.

A few things need attention before merge:

1. `@mastra/redis-streams` lint is currently failing.
   - `pubsub/redis-streams/src/agent-thread-cross-process.test.ts:19` needs a top-level type-only import for `ManagedProcess`.
   - `pubsub/redis-streams/test-fixtures/agent-thread-worker.entry.ts:23` needs `Agent` imported with `import type`.

2. The PR description appears to overstate the AgentChannels changes. It says `AgentChannels` skips `consumeStream()` when `ownerStream` is `undefined`, but this diff does not modify `packages/core/src/channels/agent-channels.ts`; current channel routing still ignores the return from `sendMessage`. If that integration is intentionally stacked separately, please update the description to make that clear.

3. Please double-check the lease-loser lifecycle. In the loser path, `signal-enqueued` is published through fire-and-forget `#publish()` before `ownerStream` resolves to `undefined`. If serverless callers use `ownerStream` as the “safe to exit” boundary for losers, consider awaiting `#publishAndWait()` there so the enqueue is flushed before the loser resolves.

Validation I ran:

- `pnpm build:core` passed from cache.
- `pnpm --filter ./packages/core check` passed.
- Focused core Vitest for `agent-signals` and `event-emitter` passed: 101 tests.
- `pnpm --filter @mastra/redis-streams build` passed.
- `pnpm --filter ./packages/core lint` passed.
- `pnpm --filter @mastra/redis-streams lint` failed with the two import-style errors above.
- Redis tests were blocked locally because Docker daemon was not running, so the package `pretest` could not start Redis.

File: packages/core/src/agent/thread-stream-runtime.ts [hunk 1]
```diff
@@ -1396,27 +1467,75 @@ export class AgentThreadStreamRuntime {
...
+    const reservedKey = key;
+    const reservedRunId = runId;
+    const resolvedPubSub = this.#getPubSub(pubsub);
+    // First acquire the cross-process lease via pubsub; on win, kick off the stream. On
+    // loss, hand the user signal off to the winning process via signal-enqueued and
+    // resolve ownerStream to undefined so the caller skips local consumption.
+    const ownerStream: Promise<MastraModelOutput<OUTPUT> | undefined> = (async () => {
+      const lease = await resolvedPubSub
+        .acquireLease(reservedKey, reservedRunId, AGENT_THREAD_LEASE_TTL_MS)
+        .catch(() => ({ acquired: true as boolean, owner: reservedRunId as string | undefined }));
```
[the remark this answers is not in the document]
Comment (CalebBarnes) [reply packages/core/src/agent/thread-stream-runtime.ts]: Intentional and now documented inline (thread-stream-runtime.ts:1530-1536). Failing open on a pubsub error is deliberate: failing closed would silently drop user signals on any transient Redis blip, which is the worse failure mode for our use case. The duplicate-stream window is bounded by the lease TTL + renewal (15s) and the next clean `acquireLease` re-serializes callers. Keeping the availability-over-strict-exclusivity tradeoff.

**Comment by CalebBarnes** [comment]
@coderabbitai review

**Comment by CalebBarnes** [comment]
@coderabbitai full review

File: docs/src/content/en/reference/agents/agent.mdx [hunk 2]
```diff
@@ -653,7 +666,7 @@ const result = await agent.sendNotificationSignal(
...
...ecord: NotificationRecord, decision: NotificationDeliveryDecision, runId?: string, signal?: CreatedAgentSignal, persisted?: Promise<void>, outcome?: Promise<SendAgentSignalAccepted> }`. Here `accepted` is a boolean reporting whether the notification was accepted for delivery (distinct from the signal-routing `accepted` promise on `sendSignal`). `record` is the stored inbox record. `decision` is the delivery-policy result. `signal` and `runId` are present when ingress emits a signal immediately, including the immediate summary emitted for active high-priority notifications. `persisted` is present when the emitted signal is persisted without waking an idle thread. `outcome` is present when a signal is emitted and resolves at decision-time, once Mastra decides what to do with it: `{ action: 'wake', runId, output }` when this process runs the agent (it started or won the lease to start the run), `{ action: 'deliver', runId }` when the signal is forwarded onto an existing run, or `{ action: 'persist' }` / `{ action: 'discard' }` when nothing ran. `runId` on the outcome is present only on `wake` and `deliver`. On the `wake` action, `output` is the agent stream for in-process consumption.
```
[the remark this answers is not in the document]
Comment (CalebBarnes) [reply docs/src/content/en/reference/agents/agent.mdx]: Fixed in fc6efb8643 — updated the `sendNotificationSignal()` return contract to the renamed shape: dropped the vestigial `accepted` boolean and renamed `outcome` -> `accepted?: Promise<SendAgentSignalAccepted>`.

File: packages/core/src/agent/agent.ts [hunk 3]
```diff
@@ -7205,44 +7221,57 @@ export class Agent<
...
           });
-          results.push({ ...result, record: summarized, decision });
+          results.push({
+            record: summarized,
+            decision,
+            runId: 'runId' in summaryAccepted ? summaryAccepted.runId : undefined,
+            signal: result.signal,
+            persisted: result.persisted,
+            accepted: result.accepted,
+          });
```
[the remark this answers is not in the document]
Comment (CalebBarnes) [reply packages/core/src/agent/agent.ts:7269]: Good catch — fixed in fc6efb8643. Both the summary and delivery paths now branch on `accepted.action` and skip the `delivered`/`summarySignalId` write when it resolves to `persist`/`discard`, pushing the record unchanged with the signal/promise attached.

File: packages/core/src/agent/thread-stream-runtime.ts [hunk 4]
```diff
@@ -89,6 +103,13 @@ type AgentThreadRuntimeState = {
   watchedThreadRunIds: Set<string>;
   preparedRunsById: Map<string, PreparedThreadRun>;
   abortedRunIds: Set<string>;
+  /**
+   * Active lease-renewal timers keyed by runId. Set when the owner
+   * process wins the cross-process lease, cleared on release. Stored
+   * here (not on a Map<key,timer>) so a run's renewal timer survives even
+   * if `activeThreadRunIds` is rotated by a follow-up signal.
+   */
+  leaseRenewalTimers: Map<string, ReturnType<typeof setInterval>>;
```
[the remark this answers is not in the document]
Comment (CalebBarnes) [reply packages/core/src/agent/thread-stream-runtime.ts:126]: Not applicable — there is no reset/dispose path that drops runtime state. State lives in a `WeakMap<PubSub, state>` (no `clearState`/`reset` method exists), so there is no code path that discards state while leaving a renewal timer running. The timers are `unref()`'d and self-clear: each run's timer is cleared in `#stopLeaseRenewal` (called from `#releaseThreadLease` on run completion/abort/fail), and a renewal that observes a lost lease stops itself. Skipping.

File: packages/core/src/agent/thread-stream-runtime.ts [hunk 5]
```diff
@@ -564,6 +665,7 @@ export class AgentThreadStreamRuntime {
       if (state.activeThreadRunIds.get(key) === record.runId) {
         state.activeThreadRunIds.delete(key);
       }
+      this.#releaseThreadLease(pubsub, key, record.runId);
       this.#publish(pubsub, key, { type: 'run-completed', runId: record.runId });
       void this.#drainPendingSignals(state, pubsub, key, record);
```
[the remark this answers is not in the document]
Comment (CalebBarnes) [reply packages/core/src/agent/thread-stream-runtime.ts]: Skipping — this ordering is intentional and correct. Line 668 releases the lease for the run that just *finished*; the successor started by `#drainPendingSignals` (line 670) issues a fresh `agent.stream(...)` that re-enters the wake path and acquires its own lease independently. Holding the finished run's lease across a new run would be the bug. The release and the successor's acquire are decoupled by design.

File: packages/core/src/harness/harness.ts [hunk 6]
```diff
@@ -1966,7 +1967,9 @@ export class Harness<TState = {}> {
           threadId,
           ifIdle: { streamOptions: streamOptions as any },
         });
-        this.emit({ type: 'follow_up_queued', count: this.#session.followUps.count(), runId: result.runId });
+        const accepted = await result.accepted.catch(() => undefined);
+        const runId = accepted && 'runId' in accepted ? accepted.runId : undefined;
+        this.emit({ type: 'follow_up_queued', count: this.#session.followUps.count(), runId });
```
[the remark this answers is not in the document]
Comment (CalebBarnes) [reply packages/core/src/harness/harness.ts]: Agreed — fixed in fc6efb8643. Dropped the `.catch(() => undefined)`. Since `next` is already dequeued, a rejected `accepted` now propagates to the outer catch so the follow-up is requeued instead of being swallowed into a false success.

File: packages/server/src/server/handlers/agents.ts [hunk 7]
```diff
@@ -1789,11 +1818,16 @@ export const SEND_AGENT_SIGNAL_ROUTE: ServerRoute<
...
+      // `accepted` carries the authoritative `runId` for `wake`/`deliver` (a run exists).
+      // `persist`/`discard` never start a run; the stored-message id (`result.signal.id`)
+      // is the correlatable id for those, keeping the wire contract (`runId: string`) stable.
+      const settled = await result.accepted;
+      const settledRunId = 'runId' in settled ? settled.runId : result.signal.id;
       return result.signal === undefined
-        ? { accepted: result.accepted, runId: result.runId }
-        : { accepted: result.accepted, runId: result.runId, signal: result.signal };
+        ? { accepted: true as const, runId: settledRunId }
+        : { accepted: true as const, runId: settledRunId, signal: result.signal };
```
[the remark this answers is not in the document]
Comment (CalebBarnes) [reply packages/server/src/server/handlers/agents.ts:1828]: Fixed in fc6efb8643 — the signal route now uses `result.signal?.id`, matching the message route. Note: at runtime the `persist`/`discard` paths always carry `signal`, so this is defensive hardening / type-level correctness rather than a live deref, but the optional chaining is the right call.

File: pubsub/redis-streams/src/index.ts [hunk 8]
```diff
@@ -305,6 +305,84 @@ export class RedisStreamsPubSub extends PubSub {
...
+    const result = await this.#writeClient.set(redisKey, owner, { NX: true, PX: ttlMs });
+    if (result === 'OK') return { acquired: true, owner };
+    // Someone holds the key — check if it's us, and renew if so.
+    const current = await this.#writeClient.get(redisKey);
+    if (current === owner) {
+      // Same owner re-claiming: refresh TTL.
+      await this.#writeClient.pExpire(redisKey, ttlMs);
+      return { acquired: true, owner };
+    }
+    return { acquired: false, owner: current ?? undefined };
```
[the remark this answers is not in the document]
Comment (CalebBarnes) [reply pubsub/redis-streams/src/index.ts]: Real bug — fixed in fc6efb8643. The re-acquire path now uses a Lua-guarded owner-checked `PEXPIRE` (GET == owner then PEXPIRE in one `EVAL`), so we can no longer extend a *new* owner's lease if the key expired and was re-claimed between the GET and the TTL refresh. This now matches the atomicity of `releaseLease`/`renewLease`.

**Comment by CalebBarnes** [comment]
@coderabbitai full review

**Comment by TylerBarnes** [comment]
One remaining correctness concern: follow-up runs started from queued work don’t appear to acquire a fresh lease.

After a run completes, `#watchThreadRunCompletion` releases the lease and then drains queued work. The drain paths (`#drainPendingSignals`, `#drainPendingIdleSignals`, `#startContinuation`) can call `agent.stream(...)` for the next run, but I don’t see a new lease acquire/renew before that stream starts.

So `queueMessage()` or other queued-after-completion work can start run 2 locally while a fresh process can also acquire the now-free Redis lease and start another idle run for the same thread.

Could you protect drained follow-up runs with their own lease before starting `agent.stream(...)`? A Redis cross-process test where a fresh worker races while queued follow-up work is draining would cover this.

**Comment by CalebBarnes** [comment]
Fixed in 5cfd24815f. The runtime no longer releases the thread lease between a completed run and its queued follow-up work — it now **holds the lease across the entire drain chain** and atomically **transfers** it to the next run's id (`transferLease`, owner-verified Lua `GET == fromOwner → SET toOwner PX ttl` in Redis) instead of release-then-acquire. The lease key is never empty while queued work exists, so a fresh process can't win the freed lease and start a competing run for the same thread.

Every drain path (`#drainPendingSignals`, `#drainPendingIdleSignals`, continuations) now acquires-or-transfers the lease before streaming, or forwards the pending signal to the current owner via `signal-enqueued` if it can't obtain it — so subscriber-side drains on non-owning processes can't start competing runs either. Providers without `transferLease` degrade to release-then-acquire (Redis is the only cross-process implementor and is gap-free).

Added a cross-process regression test that polls the raw Redis lease key directly (an earlier event-based attempt was a false negative — subscriber `run-registered` events populate local `activeThreadRunIds` and mask the freed key). It proves the lease key never goes empty during the drain and a racing process can't win a competing run. Also made the lease TTL/renewal overridable via env vars to keep these tests fast.

**Comment by CalebBarnes** [comment]
Follow-up tweak in 84035237df: made `LeaseProvider.transferLease` **required** instead of optional. Previously it was optional with a caller-side release+acquire fallback in the thread runtime's `#transferThreadLease`. The cleaner ownership is to push that decision down into each backend — every backend can implement `transferLease` (atomic backends like Redis do it gap-free; non-atomic backends implement it as `releaseLease` + `acquireLease` internally and document that they own that race window). 

Net effect: the runtime now has a single code path (no `typeof leaseProvider.transferLease === 'function'` branch), and atomicity is an explicit per-backend guarantee rather than a silent caller fallback. Also added a `transferLease` impl to `EventEmitterPubSub` (atomic in-memory owner-guarded swap) so the in-process backend is gap-free too, and `isLeaseProvider` now checks for `transferLease`. No released version exposes the optional shape, so there's no break.

**Review by TylerBarnes — APPROVED** [review]
